File: py_payment_api/app/verifiers/woocommerce_verifier.py
"""WooCommerce webhook signature verifier."""
import hmac
import hashlib
import requests
import os
from typing import Dict, Any, Optional
from .abstract.signature_verifier import SignatureVerifier
from services.logger import logger

class WooCommerceVerificationError(Exception):
    """Base exception for WooCommerce verification errors."""
    pass

class WooCommerceVerifier(SignatureVerifier):
    """
    Verifier class for WooCommerce webhook signatures and API integration.
    
    This class handles all WooCommerce-related webhook verification and API requests.
    
    API Docs: https://woocommerce.github.io/woocommerce-rest-api-docs/
    """
    
    def __init__(self) -> None:
        """
        Initialize the verifier with configuration from environment variables.
        
        Required environment variables:
        - WOOCOMMERCE_WEBHOOK_SECRET: The webhook shared secret for signature verification
        - WOOCOMMERCE_CONSUMER_KEY: The consumer key for API authentication
        - WOOCOMMERCE_CONSUMER_SECRET: The consumer secret for API authentication
        - WOOCOMMERCE_API_URL: The WooCommerce store URL
        """
        super().__init__('WOOCOMMERCE_WEBHOOK_SECRET')
        
        self._consumer_key = os.getenv('WOOCOMMERCE_CONSUMER_KEY')
        self._consumer_secret = os.getenv('WOOCOMMERCE_CONSUMER_SECRET')
        self._api_url = os.getenv('WOOCOMMERCE_API_URL')
        
        if not all([self._consumer_key, self._consumer_secret, self._api_url]):
            raise ValueError("Missing required WooCommerce configuration")

    def verify_signature(self, data: Dict[str, Any], signature: str) -> bool:
        """
        Verify the webhook signature using HMAC-SHA256.
        
        Parameters
        ----------
        data : Dict[str, Any]
            The webhook payload
        signature : str
            The signature to verify
            
        Returns
        -------
        bool
            True if signature is valid, False otherwise
        """
        try:
            # Create HMAC signature using webhook secret
            payload = str(data).encode('utf-8')
            computed = hmac.new(
                self._secret.encode('utf-8'),
                payload,
                hashlib.sha256
            ).hexdigest()
            
            # Use constant-time comparison
            return hmac.compare_digest(computed, signature)
        except Exception as e:
            logger.error(f"WooCommerce signature verification error: {str(e)}")
            return False

    def get_signature_from_header(self, header: Dict[str, Any]) -> str:
        """
        Extract the signature from the header.
        
        Parameters
        ----------
        header : Dict[str, Any]
            The header containing the signature
            
        Returns
        -------
        str
            The extracted signature
            
        Raises
        ------
        ValueError
            If signature is missing or invalid
        """
        signature = header.get('X-WC-Webhook-Signature')
        if not signature:
            raise ValueError("Missing WooCommerce signature in headers")
        if not isinstance(signature, str):
            raise ValueError("Invalid WooCommerce signature format")
        return signature

    def verify_order(self, order_id: str, order_key: Optional[str] = None) -> Dict[str, Any]:
        """
        Verify an order using the WooCommerce REST API.
        
        Parameters
        ----------
        order_id : str
            The order ID to verify
        order_key : str, optional
            The order key for additional verification
            
        Returns
        -------
        Dict[str, Any]
            The verification result containing order details
            
        Raises
        ------
        WooCommerceVerificationError
            If verification fails
        """
        try:
            # Build API endpoint URL
            endpoint = f"{self._api_url}/wp-json/wc/v3/orders/{order_id}"
            
            # Add order key if provided
            params = {}
            if order_key:
                params['order_key'] = order_key
            
            # Make API request with authentication
            response = requests.get(
                endpoint,
                auth=(self._consumer_key, self._consumer_secret),
                params=params
            )
            
            if response.status_code != 200:
                raise WooCommerceVerificationError(
                    f"Order verification failed: {response.text}"
                )
            
            return {
                'status': 'success',
                'order': response.json()
            }
            
        except Exception as e:
            logger.error(f"WooCommerce order verification error: {str(e)}")
            return {
                'status': 'error',
                'error': {
                    'msg': str(e)
                }
            }

    def verify_subscription(self, subscription_id: str) -> Dict[str, Any]:
        """
        Verify a subscription using the WooCommerce Subscriptions API.
        
        Parameters
        ----------
        subscription_id : str
            The subscription ID to verify
            
        Returns
        -------
        Dict[str, Any]
            The verification result containing subscription details
            
        Raises
        ------
        WooCommerceVerificationError
            If verification fails
        """
        try:
            endpoint = f"{self._api_url}/wp-json/wc/v3/subscriptions/{subscription_id}"
            
            response = requests.get(
                endpoint,
                auth=(self._consumer_key, self._consumer_secret)
            )
            
            if response.status_code != 200:
                raise WooCommerceVerificationError(
                    f"Subscription verification failed: {response.text}"
                )
            
            return {
                'status': 'success',
                'subscription': response.json()
            }
            
        except Exception as e:
            logger.error(f"WooCommerce subscription verification error: {str(e)}")
            return {
                'status': 'error',
                'error': {
                    'msg': str(e)
                }
            }

    def handle_notification(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """
        Handle different types of WooCommerce webhook notifications.
        
        Parameters
        ----------
        payload : Dict[str, Any]
            The webhook payload
            
        Returns
        -------
        Dict[str, Any]
            The processed notification result
        """
        try:
            event = payload.get('webhook_event')
            
            if event in ['order.created', 'order.updated']:
                return self._handle_order_notification(payload)
            elif event in ['subscription.created', 'subscription.renewed']:
                return self._handle_subscription_notification(payload)
            elif event == 'subscription.cancelled':
                return self._handle_subscription_cancellation(payload)
            
            return {
                'status': 'error',
                'type': 'unknown_event',
                'update_body': {},
                'identifier': None
            }
            
        except Exception as e:
            logger.error(f"WooCommerce notification handling error: {str(e)}")
            return {
                'status': 'error',
                'type': 'processing_error',
                'update_body': {'error': str(e)},
                'identifier': None
            }

    def _handle_order_notification(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle order-related notifications."""
        return {
            'status': 'action_update',
            'type': 'order_update',
            'update_body': {
                'status': payload.get('status'),
                'order_id': payload.get('id')
            },
            'identifier': {
                'order_id': payload.get('id'),
                'product_id': payload.get('line_items', [{}])[0].get('product_id')
            }
        }

    def _handle_subscription_notification(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle subscription-related notifications."""
        return {
            'status': 'action_update',
            'type': 'subscription_update',
            'update_body': {
                'status': 'active',
                'subscription_id': payload.get('id')
            },
            'identifier': {
                'subscription_id': payload.get('id'),
                'product_id': payload.get('line_items', [{}])[0].get('product_id')
            }
        }

    def _handle_subscription_cancellation(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle subscription cancellation notifications."""
        return {
            'status': 'action_update',
            'type': 'subscription_cancelled',
            'update_body': {
                'status': 'cancelled',
                'subscription_id': payload.get('id')
            },
            'identifier': {
                'subscription_id': payload.get('id'),
                'product_id': payload.get('line_items', [{}])[0].get('product_id')
            }
        } 

File: tsPaymentApi/src/verifiers/woocommerceVerifier.ts
import { SignatureVerifier } from './abstract/signatureVerifier';
import axios from 'axios';
import crypto from 'crypto';
import { logger } from '../services/logger';

export class WooCommerceVerificationError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'WooCommerceVerificationError';
    }
}

interface WooCommerceConfig {
    consumerKey: string;
    consumerSecret: string;
    webhookSecret: string;
    apiUrl: string;
}

interface VerifyResponse {
    status: 'success' | 'error';
    order?: any;
    subscription?: any;
    error?: {
        msg: string;
    };
}

interface NotificationResponse {
    status: 'action_update' | 'error';
    type: string;
    update_body: Record<string, any>;
    identifier: Record<string, any> | null;
}

export class WooCommerceVerifier extends SignatureVerifier {
    private config: WooCommerceConfig;

    constructor() {
        super('WOOCOMMERCE_WEBHOOK_SECRET');
        
        this.config = {
            consumerKey: process.env.WOOCOMMERCE_CONSUMER_KEY!,
            consumerSecret: process.env.WOOCOMMERCE_CONSUMER_SECRET!,
            webhookSecret: process.env.WOOCOMMERCE_WEBHOOK_SECRET!,
            apiUrl: process.env.WOOCOMMERCE_API_URL!
        };

        if (!this.config.consumerKey || !this.config.consumerSecret || !this.config.webhookSecret) {
            throw new Error('Missing required WooCommerce configuration');
        }
    }

    public async verifySignature(payload: string, signature: string): Promise<boolean> {
        try {
            const hmac = crypto.createHmac('sha256', this.config.webhookSecret);
            const calculatedSignature = hmac.update(payload).digest('hex');
            return crypto.timingSafeEqual(
                Buffer.from(calculatedSignature),
                Buffer.from(signature)
            );
        } catch (error) {
            logger.error('WooCommerce signature verification error:', error);
            return false;
        }
    }

    public async verifyOrder(orderId: string, orderKey?: string): Promise<VerifyResponse> {
        try {
            const endpoint = `${this.config.apiUrl}/wp-json/wc/v3/orders/${orderId}`;
            const params = orderKey ? { order_key: orderKey } : {};

            const response = await axios.get(endpoint, {
                auth: {
                    username: this.config.consumerKey,
                    password: this.config.consumerSecret
                },
                params
            });

            return {
                status: 'success',
                order: response.data
            };

        } catch (error) {
            logger.error('WooCommerce order verification error:', error);
            return {
                status: 'error',
                error: {
                    msg: error instanceof Error ? error.message : 'Unknown error'
                }
            };
        }
    }

    public async verifySubscription(subscriptionId: string): Promise<VerifyResponse> {
        try {
            const endpoint = `${this.config.apiUrl}/wp-json/wc/v3/subscriptions/${subscriptionId}`;

            const response = await axios.get(endpoint, {
                auth: {
                    username: this.config.consumerKey,
                    password: this.config.consumerSecret
                }
            });

            return {
                status: 'success',
                subscription: response.data
            };

        } catch (error) {
            logger.error('WooCommerce subscription verification error:', error);
            return {
                status: 'error',
                error: {
                    msg: error instanceof Error ? error.message : 'Unknown error'
                }
            };
        }
    }

    public async handleNotification(payload: Record<string, any>): Promise<NotificationResponse> {
        try {
            const event = payload.webhook_event as string;
            
            if (event.startsWith('order.')) {
                return this.handleOrderNotification(payload);
            } else if (event.startsWith('subscription.')) {
                return event === 'subscription.cancelled' 
                    ? this.handleSubscriptionCancellation(payload)
                    : this.handleSubscriptionNotification(payload);
            }
            
            return {
                status: 'error',
                type: 'unknown_event',
                update_body: {},
                identifier: null
            };
            
        } catch (error) {
            logger.error('WooCommerce notification handling error:', error);
            return {
                status: 'error',
                type: 'processing_error',
                update_body: { error: error instanceof Error ? error.message : 'Unknown error' },
                identifier: null
            };
        }
    }

    private handleOrderNotification(payload: Record<string, any>): NotificationResponse {
        return {
            status: 'action_update',
            type: 'order_update',
            update_body: {
                status: payload.status,
                order_id: payload.id
            },
            identifier: {
                order_id: payload.id,
                product_id: payload.line_items?.[0]?.product_id
            }
        };
    }

    private handleSubscriptionNotification(payload: Record<string, any>): NotificationResponse {
        return {
            status: 'action_update',
            type: 'subscription_update',
            update_body: {
                status: 'active',
                subscription_id: payload.id
            },
            identifier: {
                subscription_id: payload.id,
                product_id: payload.line_items?.[0]?.product_id
            }
        };
    }

    private handleSubscriptionCancellation(payload: Record<string, any>): NotificationResponse {
        return {
            status: 'action_update',
            type: 'subscription_cancelled',
            update_body: {
                status: 'cancelled',
                subscription_id: payload.id
            },
            identifier: {
                subscription_id: payload.id,
                product_id: payload.line_items?.[0]?.product_id
            }
        };
    }
}

export const woocommerceVerifier = new WooCommerceVerifier(); 